pytorch_lightning/trainer/trainer.py

```diff
@@ -853,11 +853,11 @@ def _run(self, model: 'pl.LightningModule') -> Optional[Union[_EVALUATE_OUTPUT,
         self.accelerator.setup_environment()
         self._call_setup_hook(model)  # allow user to setup lightning_module in accelerator environment
 
+        # restore callback states
+        self.checkpoint_connector.restore_callbacks()
         # restore modules after setup
         self.checkpoint_connector.restore_datamodule()
         self.checkpoint_connector.restore_model()
-        # restore callback states
-        self.checkpoint_connector.restore_callbacks()
 
```

Comment:
is it necessary? this will affect all callbacks. maybe it is possible to implement the QAT callback using the setup() method which runs right before restore. What do you think?

Reply:
> maybe it is possible to implement the QAT callback using the setup() method which runs right before restore

which setup, setup of a model? but the model does not know anything about callback which is added for training
if you write a model which counts on the run only with QAT it can work but how about models which are kind of vanilla and later you just want to train with QAT?

as we spoke earlier, the actual calling `on_load_checkpoint` after the model is restored is not logical and it shall be before... so what I need to have an entry with access to checkpoint before te model is restored and it shall be independent on the model code :rabbit:

Reply:
then I would like to suggest adding a changelog entry under "changed" because we won't be able to make this backward compatible. 

To the best of my knowledge, the order was always

```
DataModule.on_load_checkpoint
Model.on_load_checkpoint
Callbacks.on_load_checkpoint
```

Reply:
> To the best of my knowledge, the order was always

but in such a case when a callback can be restored and do some model changes before restoring?
On the other hand, the Trainer itself is restored before, so it is:
```
Trainer.restore...
# Callbacks.on_load_checkpoint # suggestions
DataModule.on_load_checkpoint
Model.on_load_checkpoint
Callbacks.on_load_checkpoint # actual now
```

Reply:
> On the other hand, the Trainer itself is restored before, so it is:

no, this is not the case. The trainer, in terms of loop state, optimizers etc. gets restored last, just before training begins. 

the `on_load_checkpoint` hooks are the very first hooks that run while restoring so the user has a chance to modify the checkpoint dict.

Reply:
I would be very careful about changing the order.

In master, callbacks can expect the `LightningModule` to be fully loaded before loading themselves, one example is the `BaseFinetuningCallback`.

https://github.com/PyTorchLightning/pytorch-lightning/blob/b7dbcc3e13a2e34e9fc1850d461909caf8ab11e0/pytorch_lightning/callbacks/finetuning.py#L95-L103

Changing this could break many existing callbacks.

I am not very familiar with what you are trying to fix. Can you elaborate on why do you need to `self.prepare_model(pl_module)` before the model is loaded?

Reply:
@tchaton is fixing finetuning.py

Reply:
> Can you elaborate on why do you need to `self.prepare_model(pl_module)` before the model is loaded?

I need to add some states to the modules which are used for collecting stats, otherwise, the loading fails as checkpointed model contains attributes that are not presented in the vanilla model

Reply:
But why don't you remove any quantization related data on the model before the model gets checkpointed?

Then loading it wouldn't fail without the callback

Any of the callback logic should be entirely saved by the callback, and when loaded, re-applied to the model.

Reply:
as poke with @t-vi of we drop all related attributes that resume does not make sense as you removed them...

> Any of the callback logic should be entirely saved by the callback, and when loaded, re-applied to the model.

so I will save the full model again as a callback state, that is what you suggest?